pallets/fees/src/migration.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+// This file is part of Centrifuge (centrifuge.io) parachain.
+
+// Cumulus is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+
+use codec::{Decode, Encode};
+use frame_support::{log, traits::Get, weights::Weight, Blake2_256};
+
+#[cfg(feature = "try-runtime")]
+use frame_support::ensure; // Not in prelude for try-runtime
+
+#[derive(Encode, Decode, Default, Clone, PartialEq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Debug))]
+struct Fee<Hash, Balance> {
+	key: Hash,
+	price: Balance,
+}
+
+#[frame_support::storage_alias]
+type Fees<T: Config> = StorageMap<
+	_,
+	Blake2_256,
+	<T as frame_system::Config>::Hash,
+	Fee<<T as frame_system::Config>::Hash, BalanceOf<T>>,
+>;
+
+pub mod fee_balances {
+	use super::*;
+
+	#[cfg(feature = "try-runtime")]
+	pub fn pre_migrate<T: Config>() -> Result<(), &'static str> {
+		let count = Fees::<T>::iter_values().count();
+		ensure!(count != 0, "Fees storage items not found!");
+		log::info!(
+				target: "runtime::fees::pre-migrate",
+				"Pre Migrate check passed with count {:?}",
+				count,
+		);
+		Ok(())
+	}
+
+	pub fn migrate<T: Config>() -> Weight {
```

Comment:
I noticed that since the migration execution depends on each pallet hook implementation and is not easily visible anymore in the runtime file, then is easy to forget that the migration is still there once it has been executed. (Noticed with the Claims migration)
So this means that if we run a migration twice (accidentally) and the second time it fails then we can incur in a chain halt (worst case scenario)
In this scenario, I think the migrate code will delete every entry in the fees storage so I think we should add here a check that do not run the migration twice.
I think doing something like:
```
let on_chain_storage_version = <P as GetStorageVersion>::on_chain_storage_version();
if on_chain_storage_version < 1 {
  // run migration
  // StorageVersion::new(1).put::<P>();
} else {
  // skip with warn message0
}
```
Maybe there is a better way to do this? WDYT?

Reply:
The `migration()` function is only removing the old storage and does nothing with the new (and valid) one. So if the migration is run twice, the second one no has an effect, you would try to remove the already removed storage. Please, tell me if I'm misunderstanding this and if there is some issue under the hood that I'm not seeing.

Reply:
Ah! For some reason I thought that the storage name/prefix was the same. 👍